File: android/app/src/main/java/com/engagendy/noor/ReaderChrome.kt
package com.engagendy.noor

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/// Shared reader-chrome state — the Android twin of iOS `ReaderChrome`
/// (`Modules/QuranReader/Sources/QuranReader/ReaderChrome.swift`).
///
/// Immersive reading: while a Quran reader is open the bottom tab bar
/// follows the reader's CHROME rather than the whole session, exactly as on
/// iOS — tap the page and the top strip AND the tab bar come back together,
/// tap again and both go.
///
/// This is the single source of truth: `MushafScreen` owns `chromeVisible`
/// through this object (it holds NO private copy), so the strip and the bar
/// can never disagree. `QuranScreen` raises/lowers `readerOpen`;
/// `MainActivity` also gates on the Quran tab, so the bar can never be
/// stranded off-screen after a tab switch.
///
/// The tab bar must never RESIZE the reader when it returns: a Madani page is
/// a rigid 15-row grid stretched to the height it gets, so every line would
/// reflow on every tap. `MainActivity` therefore draws it as an overlay above
/// the reader (never in the Scaffold's `bottomBar`, which insets content by
/// design) while the reader keeps the full height.
object ReaderChrome {
    /// True while a Quran reader (Madani page, flow, or ayah-by-ayah) is on
    /// screen.
    var readerOpen by mutableStateOf(false)

    /// True while the reader's chrome (top strip) is showing. Backed by a
    /// `MutableState` so the Madani reader can delegate to it with `by`
    /// instead of keeping — and having to mirror — a local copy.
    ///
    /// Only the Madani reader hides its chrome; the flow / ayah-by-ayah
    /// reader's top bar is permanently visible, so it simply leaves this
    /// `true` and keeps its tab bar.
    val chromeState = mutableStateOf(true)
    var chromeVisible: Boolean
        get() = chromeState.value
        set(value) { chromeState.value = value }

    /// True while the open reader is one whose chrome HIDES — today only the
    /// Madani page reader. Its bar must float over the page (a rigid 15-row
    /// grid would reflow if the bar took height); the flow / ayah-by-ayah
    /// reader keeps a permanently visible top bar, so it keeps a normal
    /// inline tab bar that insets its scrolling text instead of covering it.
    var chromeHides by mutableStateOf(false)

    /// One animation for the strip and the bar, so a tap reads as a single
    /// gesture (the strip cross-fades with `animateFloatAsState`, which uses
    /// exactly this spring by default).
    val fadeSpec = spring<Float>(
        dampingRatio = Spring.DampingRatioNoBouncy,
        stiffness = Spring.StiffnessMedium,
        visibilityThreshold = 0.01f)

    /// Reader appeared: chrome starts visible (the Madani reader auto-hides
    /// it shortly after arrival).
    fun readerAppeared() {
        readerOpen = true
        chromeVisible = true
    }

    /// Reader left by ANY path — drawer exit row, tab switch, system back.
    fun readerDisappeared() {
        readerOpen = false
        chromeVisible = true
        chromeHides = false
    }
}

File: Modules/QuranReader/Sources/QuranReader/ReaderChrome.swift
import SwiftUI
#if canImport(UIKit)
import UIKit
#endif

/// Shared reader-chrome state — the iOS twin of Android's `ReaderChrome`.
///
/// The reader is immersive: the system tab bar is hidden for the whole
/// session. But iOS has no system back button, and with the navigation bar
/// hidden the interactive edge-swipe back is gone too (verified on the
/// simulator), so in the fully immersive state the drawer's "Back to Quran"
/// row is the only way out. To give a familiar escape, the app layer draws a
/// floating tab bar over the reader whenever the reader's chrome is up: tap
/// the page, the top strip AND the bar fade in together; tap again, both go.
///
/// This is the single source of truth for that — `SurahReaderView` owns
/// `chromeVisible` through this object (it has no private copy), so the bar
/// can never disagree with the strip.
///
/// Deliberate divergence from Android, which keeps its bar hidden for the
/// whole reader session because it has a system back button. Do not
/// "fix" the two back into symmetry.
@MainActor
@Observable
public final class ReaderChrome {
    public static let shared = ReaderChrome()

    /// True while a Quran reader is on screen.
    public private(set) var readerOpen = false
    /// True while the reader's chrome (top strip) is showing.
    public var chromeVisible = true

    private init() {}

    /// Reader appeared: chrome starts visible (it auto-hides shortly after).
    public func readerAppeared() {
        readerOpen = true
        chromeVisible = true
    }

    /// Reader left by ANY path — drawer exit row, tab switch, pop.
    public func readerDisappeared() {
        readerOpen = false
        chromeVisible = false
    }

    /// Height of the home indicator (0 on a home-button device).
    ///
    /// The reader ignores its container's bottom safe area so the tab bar
    /// floats over the page instead of resizing it, then pads this back in
    /// to keep exactly the layout it had before the bar existed. It must
    /// come from the WINDOW: a `GeometryProxy`'s bottom inset includes the
    /// tab bar — the very thing being ignored — while the window's carries
    /// only the system inset and never moves when the bar appears.
    public static var homeIndicatorInset: CGFloat {
        #if canImport(UIKit)
        UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow }
            .first?.safeAreaInsets.bottom ?? 0
        #else
        0
        #endif
    }
}

